File: backend/src/models/AuditRecord.ts
import mongoose, { Schema, type Document } from 'mongoose';

export interface IAuditRecord extends Document {
  actorId?: mongoose.Types.ObjectId;
  actorRole?: string;
  action: string;
  entityType?: string;
  entityId?: mongoose.Types.ObjectId | string;
  metadata?: Record<string, unknown>;
  ipAddress?: string;
  createdAt: Date;
}

const auditRecordSchema = new Schema<IAuditRecord>(
  {
    actorId: { type: Schema.Types.ObjectId, ref: 'User' },
    actorRole: { type: String, trim: true },
    action: { type: String, required: true, trim: true },
    entityType: { type: String, trim: true },
    entityId: { type: Schema.Types.Mixed },
    metadata: { type: Schema.Types.Mixed, default: {} },
    ipAddress: { type: String, trim: true }
  },
  { timestamps: { createdAt: true, updatedAt: false } }
);

auditRecordSchema.index({ actorId: 1, createdAt: -1 });
auditRecordSchema.index({ entityType: 1, entityId: 1, createdAt: -1 });

export const AuditRecord = mongoose.model<IAuditRecord>('AuditRecord', auditRecordSchema);

File: backend/src/main/java/com/raava/model/AuditRecord.java
package com.raava.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.Map;

@Document(collection = "audit_records")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class AuditRecord {
    @Id
    private String id;

    @DBRef(lazy = true)
    private User actor;

    private String actorRole;
    private String action;
    private String entityType;
    private String entityId;
    private Map<String, Object> metadata;

    @Indexed
    private Instant createdAt;

    @Indexed
    private Instant updatedAt;
}
